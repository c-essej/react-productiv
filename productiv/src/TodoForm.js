import React, { useState } from "react";

const defaultFormData = {
  title: "",
  description: "",
  priority: 3
};

/** Form for adding.
 *
 * Props:
 * - initialFormData
 * - handleSave: function to call in parent.
 *
 * { TodoApp, EditableTodo } -> TodoForm
 */
//TODO: update docstring to reflect state
function TodoForm({ initialFormData=defaultFormData, handleSave }) {

  const [formData, setFormData] = useState(initialFormData);
  // console.log(`TodoForm rendering with form data`, formData);
  /** Update form input. */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(fData => ({
      ...fData,
      [name]: value,
    }));
  }

  /** Call parent function and clear form. */
  function handleSubmit(evt) {
    evt.preventDefault();
    handleSave(formData);
    setFormData(initialFormData);
  };
//TODO: give labels to input elements
  return (
    <form className="NewTodoForm" onSubmit={handleSubmit}>

      <div className="mb-3">
      <label htmlFor="newTodo-title">Title:</label>

        <input
          id="newTodo-title"
          name="title"
          className="form-control"
          placeholder="Title"
          onChange={handleChange}
          value={formData.title}
          aria-label="Title"
        />
      </div>

      <div className="mb-3">
      <label htmlFor="newTodo-description">Description:</label>
        <textarea
          id="newTodo-description"
          name="description"
          className="form-control"
          placeholder="Description"
          onChange={handleChange}
          value={formData.description}
          aria-label="Description"
        />
      </div>

      <div className="mb-3 d-flex justify-content-between">
        <div className="w-75 d-flex justify-content-between">
          <label htmlFor="newTodo-priority"
            className="d-inline-flex">Priority:</label>
          <select id="newTodo-priority"
            name="priority"
            value={formData.priority}
            onChange={handleChange}
            className="form-control form-control-sm d-inline-flex"
          >
            <option value={1}>Ultra-Über</option>
            <option value={2}>Über</option>
            <option value={3}>Meh</option>
          </select>
        </div>
        <button className="btn-primary rig btn btn-sm NewTodoForm-addBtn">
          Gø!
        </button>
      </div>

    </form>
  );
}

export default TodoForm;
