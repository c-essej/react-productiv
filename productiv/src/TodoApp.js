import React, { useState } from "react";
import { v4 as uuid } from "uuid";

import TopTodo from "./TopTodo";
import EditableTodoList from "./EditableTodoList";
import TodoForm from "./TodoForm";

/** App for managing a todo list.
 *
 * Props:
 * - initialTodos: possible array of [ todo, ... ]
 *
 * State:
 * - todos: array of [ todo, ... ]
 *
 * App -> TodoApp -> { TodoForm, EditableTodoList }
 */

function TodoApp({ initialTodos }) {
  const [todos, setTodos] = useState(initialTodos);
  console.log('TodoApp', todos);
  let top = todos.reduce(
    (acc, cur) => cur.priority < acc.priority ? cur : acc, todos[0]);

  /** add a new todo to list */
  function create(todo) {
    const newTodo = { ...todo, id: uuid() };
    setTodos(todos => [...todos, newTodo]);
  }

  /** update a todo with updatedTodo */
  function update(updatedTodo) {
    setTodos(todos => todos.map(
      todo => todo.id === updatedTodo.id ? updatedTodo : todo
    ));
  }

  /** delete a todo by id */
  function remove(id) {
    console.log("running remove");
    setTodos(todos => todos.filter(todo => todo.id !== id));
  }

  return (
    <main className="TodoApp">
      <div className="row">

        <div className="col-md-6 EditableTodoList">
          <h3>Todos</h3>
          {todos.length > 0
            ?
            <EditableTodoList todos={todos} update={update} remove={remove} />
            :
            <span className="text-muted">You have no todos.</span>}
        </div>

        <div className="col-md-6">
          {top
            ?
            <section className="mb-4">
              <h3>Top Todo</h3>
              <TopTodo todos={todos} />
            </section>
            :
            ""
          }

          <section>
            <h3 className="mb-3 TodoForm">Add NÃ¼</h3>
            <TodoForm handleSave={create} />
          </section>
        </div>

      </div>
    </main>
  );
}

export default TodoApp;