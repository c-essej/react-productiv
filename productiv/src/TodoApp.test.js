import TodoApp from './TodoApp';
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TEST_TODOS from './_testCommon';
import TodoForm from './TodoForm';


describe("Tests for Todo App", function () {

  //smoke test
  it("Renders without crashing", function () {
    const { container } = render(<TodoApp initialTodos={TEST_TODOS} />);
  });

  //snapshot test
  it("matches snapshot", function () {
    const { container } = render(
      <TodoApp initialTodos={TEST_TODOS} />);

    expect(container).toMatchSnapshot();
  });


  it("Renders all todos", function () {
    const { container } = render(
      <TodoApp initialTodos={TEST_TODOS} />);

    expect(container).toHaveTextContent('test1');
    expect(container).toHaveTextContent('test2');
    expect(container.querySelector(".EditableTodoList")).toHaveTextContent('test1');
  });

  it("Renders todo form", function () {
    const { container } = render(
      <TodoApp initialTodos={TEST_TODOS} />);

    expect(container.querySelector(".TodoForm")).toBeInTheDocument();
  });

  it("creates a new todo", function () {
    const { container, getByLabelText, debug } = render(
      <TodoApp initialTodos={TEST_TODOS} />);

    const nameField = getByLabelText("Title:");
    const descriptionField = getByLabelText("Description:");
    const priorityField = getByLabelText("Priority:");
    const goButton = container.querySelector(".NewTodoForm-addBtn");

    fireEvent.change(nameField, { target: { value: "newTitle" } });
    fireEvent.change(descriptionField, { target: { value: "newDesc" } });
    fireEvent.change(priorityField, { target: { value: 1 } });

    fireEvent.click(goButton);

    debug(nameField);

    debug(container.querySelector(".EditableTodoList"));
    expect(container.querySelector(".EditableTodoList")).toHaveTextContent('newTitle');



  });

});
