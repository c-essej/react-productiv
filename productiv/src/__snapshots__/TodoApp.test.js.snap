// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Todo App matches snapshot 1`] = `
<div>
  <main
    class="TodoApp"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 EditableTodoList"
      >
        <h3>
          Todos
        </h3>
        <div>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
              >
                <div>
                  <b>
                    test1
                  </b>
                   
                  <small>
                    Priority 1
                  </small>
                </div>
                <div>
                  <small>
                    test1 description
                  </small>
                </div>
              </div>
            </div>
          </div>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
              >
                <div>
                  <b>
                    test2
                  </b>
                   
                  <small>
                    Priority 2
                  </small>
                </div>
                <div>
                  <small>
                    test2 description
                  </small>
                </div>
              </div>
            </div>
          </div>
          <div
            class="EditableTodo"
          >
            <div
              class="mb-3"
            >
              <div
                class="float-end text-sm-end"
              >
                <button
                  class="EditableTodo-toggle btn-link btn btn-sm"
                >
                  Edit
                </button>
                <button
                  class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                >
                  Del
                </button>
              </div>
              <div
                class="Todo"
              >
                <div>
                  <b>
                    test3
                  </b>
                   
                  <small>
                    Priority 3
                  </small>
                </div>
                <div>
                  <small>
                    test3 description
                  </small>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-6"
      >
        <section
          class="mb-4"
        >
          <h3>
            Top Todo
          </h3>
          <div>
            <div
              class="Todo"
            >
              <div>
                <b>
                  test1
                </b>
                 
                <small>
                  Priority 1
                </small>
              </div>
              <div>
                <small>
                  test1 description
                </small>
              </div>
            </div>
          </div>
        </section>
        <section>
          <h3
            class="mb-3 TodoForm"
          >
            Add Nü
          </h3>
          <form
            class="NewTodoForm"
          >
            <div
              class="mb-3"
            >
              <label
                for="newTodo-title"
              >
                Title:
              </label>
              <input
                aria-label="Title"
                class="form-control"
                id="newTodo-title"
                name="title"
                placeholder="Title"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                for="newTodo-description"
              >
                Description:
              </label>
              <textarea
                aria-label="Description"
                class="form-control"
                id="newTodo-description"
                name="description"
                placeholder="Description"
              />
            </div>
            <div
              class="mb-3 d-flex justify-content-between"
            >
              <div
                class="w-75 d-flex justify-content-between"
              >
                <label
                  class="d-inline-flex"
                  for="newTodo-priority"
                >
                  Priority:
                </label>
                <select
                  class="form-control form-control-sm d-inline-flex"
                  id="newTodo-priority"
                  name="priority"
                >
                  <option
                    value="1"
                  >
                    Ultra-Über
                  </option>
                  <option
                    value="2"
                  >
                    Über
                  </option>
                  <option
                    value="3"
                  >
                    Meh
                  </option>
                </select>
              </div>
              <button
                class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
              >
                Gø!
              </button>
            </div>
          </form>
        </section>
      </div>
    </div>
  </main>
</div>
`;
